text/3714-macro-fragment-fields.md

```diff
@@ -0,0 +1,182 @@
+specifiers are more robust for these cases, but don't allow extracting
+individual pieces of the matched syntax.
+
+This RFC introduces a mechanism to use high-level fragment specifiers while
+still extracting individual pieces of the matched syntax.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When writing macros by example, and using certain high-level fragment
+specifiers, you can use the syntax `${matched_name.field_name}` to extract
+specific "fields" of the matched syntax. This allows you to use the Rust parser
+for those high-level fragments, rather than having to recreate parts of the
+Rust grammar in order to extract the specific pieces you want. Fields evaluate
+to pieces of Rust syntax, suitable for substitution into the program or passing
+to other macros for further processing.
+
+For example, the fragment `:adt` parses any abstract data type supported by
+Rust: struct, union, or enum. Given a match `$t:adt`, you can obtain the name
+of the matched type with `${t.name}`:
+
+```rust
+macro_rules! get_name {
+    ($t:adt) => { println!("{}", stringify!(${t.name})); }
+}
+
+fn main() {
+    let n1 = get_name!(struct S { field: u32 });
+    let n2 = get_name!(enum E { V1, V2 = 42, V3(u8) });
+    let n3 = get_name!(union U { u: u32, f: f32 });
+    println!("{n3}{n1}{n2}"); // prints "USE"
+}
+```
+
+An attempt to access a field that doesn't exist will produce a compilation
+error on the macro definition, whether or not the specific macro rule gets
+invoked.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Fragment fields may be used in a macro transcriber anywhere the corresponding
+fragment name could be used.
+
+Fragment fields typically follow the same rules for repetition handling as the
+corresponding fragment (e.g. being used at the same level/kind of repetition).
+However, fragment fields that contain multiple items require one additional
+level of repetition; see the `param` field of `:fn`, below.
+
+This RFC introduces the following new fragment specifiers, with specified fields:
+
+- `:fn`: A function definition (including body).
+  - `name`: The name of the function, as an `ident`.
+  - `param`: The parameters of the function, presented as though captured by a
+    level of `*` repetition. For instance, you can write `$(${f.param}),*` to
+    get a comma-separated list of parameters, or `$(other_macro!(${f.param}))*`
+    to pass each parameter to another macro.
+  - `return_type`: The return type of the function, as a `ty`. If the function
+    has no explicitly specified return type, this will be `()`, with a span of
+    the closing parenthesis for the function arguments.
+  - `body`: The body of the function, as a block (including the
+    surrounding braces).
+  - `vis`: The visibility of the function, as a `vis` (may be empty).
+- `:adt`: An ADT (struct, union, or enum).
```

Comment:
Maybe `:adt` could also have a `.vis` field, to match `:fn`? Unless `:adt` doesn't match the vis, but then that feels mismatched with `:fn`.

Reply:
Good call. I left out most things because they varied between types of ADT, but all ADTs have visibility.